### PR TITLE
add basic folder operations to tree

### PR DESCRIPTION
**Pull request by ryanfkeepers**
adds SetFolder and AddTombstone operations
to the drive delta tree.

---

#### Does this PR need a docs update or release note?

- [x] :no_entry: No

#### Type of change

- [x] :sunflower: Feature

#### Issue(s)

* #4689

#### Test Plan

- [x] :zap: Unit test

### CODE REVIEW

File: src/internal/m365/collection/drive/delta_tree.go [hunk 1]
```diff
@@ -65,16 +70,103 @@ type nodeyMcNodeFace struct {
...
+		if face.root == nil {
+			root := newNodeyMcNodeFace(nil, id, name, nil, isPackage)
+			face.root = root
+			face.folderIDToNode[id] = root
+		}
+
+		return nil
+	}
+
+	if buriedAt, alreadyBuried := face.tombstones[id]; alreadyBuried {
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/delta_tree.go:120]: I'll go get that coverage in the delta doc.

File: src/internal/m365/collection/drive/delta_tree.go [hunk 2]
```diff
@@ -65,16 +70,103 @@ type nodeyMcNodeFace struct {
...
+		return clues.NewWC(ctx, "attempted to add tombstoned folder to tree").
+			With("tombstone_path", buriedAt)
+	}
+
+	parent, ok := face.folderIDToNode[parentID]
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/delta_tree.go:125]: Sure!  Can do.

File: src/internal/m365/collection/drive/delta_tree.go [hunk 3]
```diff
@@ -65,16 +70,103 @@ type nodeyMcNodeFace struct {
...
 		items:     map[string]time.Time{},
 		isPackage: isPackage,
 	}
 }
+
+// ---------------------------------------------------------------------------
+// folder handling
+// ---------------------------------------------------------------------------
+
+func (face *folderyMcFolderFace) SetFolder(
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/delta_tree.go:91]: Yeah, I thought of that.  May revisit as implementation progresses, but wanted to avoid for now for simplicity.  To begin with I wanted to use the method signatures to showcase the exact details about the behavior being handled.

File: src/internal/m365/collection/drive/delta_tree.go [hunk 4]
```diff
@@ -65,16 +70,103 @@ type nodeyMcNodeFace struct {
...
+		return clues.NewWC(ctx, "missing tombstone folder ID")
+	}
+
+	if len(loc) == 0 {
+		return clues.NewWC(ctx, "missing tombstone location")
+	}
+
+	if _, stillKicking := face.folderIDToNode[id]; stillKicking {
+		return clues.NewWC(ctx, "attempted to tombstone a folder that exists in the tree")
+	}
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/delta_tree.go:162]: Lets discuss details out of band.  My testing doesn't reflect those results.  But you might have taken a different approach.  Would be nice to get a repro, if it's possible.

File: src/internal/m365/collection/drive/delta_tree_test.go [hunk 5]
```diff
@@ -40,17 +41,401 @@ func (suite *DeltaTreeUnitSuite) TestNewNodeyMcNodeFace() {
...
+				folderIDToNode: map[string]*nodeyMcNodeFace{},
+			},
+			id:        odConsts.RootID,
+			name:      odConsts.RootPathDir,
+			isPackage: true,
+			expectErr: assert.NoError,
+		},
+		{
+			tname: "root already exists",
+			tree: &folderyMcFolderFace{
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/delta_tree_test.go:93]: done

File: src/internal/m365/collection/drive/delta_tree_test.go [hunk 6]
```diff
@@ -40,17 +41,401 @@ func (suite *DeltaTreeUnitSuite) TestNewNodeyMcNodeFace() {
...
+
+	// relocate last child underneath top_right
+	pid = "tr"
+	set(pid, id, "bot_right", false)
+
+	nodey = tree.folderIDToNode[id]
+	require.NotNil(t, nodey)
+	assert.Len(t, tree.folderIDToNode, 5)
+	assert.Equal(t, tr, nodey.parent)
+	assert.Equal(t, nodey, tr.children[id])
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/delta_tree_test.go:337]: aye, can do

File: src/internal/m365/collection/drive/delta_tree_test.go [hunk 7]
```diff
@@ -40,17 +41,401 @@ func (suite *DeltaTreeUnitSuite) TestNewNodeyMcNodeFace() {
...
+	assert.Len(t, tl.children, 1)
+	assert.Len(t, tr.children, 1)
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/delta_tree_test.go:339]: Yup, happy to do so.

File: src/internal/m365/collection/drive/delta_tree_test.go [hunk 8]
```diff
@@ -40,17 +41,401 @@ func (suite *DeltaTreeUnitSuite) TestNewNodeyMcNodeFace() {
...
+				parent := test.tree.folderIDToNode[test.parentID]
+				assert.Equal(t, parent, result.parent)
+			}
+		})
+	}
+}
+
+// unlike the prior test, this focuses entirely on whether or not the
+// tree produced by folder additions is correct.
+func (suite *DeltaTreeUnitSuite) TestFolderyMcFolderFace_SetFolder_correctTree() {
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/delta_tree_test.go:247]: On one hand, I'm okay exposing internals like this.  It makes for strong and clear assertions about the data management of the tree across changes.

On the other hand, I was also planning on including a tree walking func when I handle the post-processing work.  Was planning to keep both tests, but I can look at shaping this up in the process for flexibility.  I think I'm going to keep it as it is for now, but rest assured I'll keep it in mind while I move forward with other changes.

File: src/internal/m365/collection/drive/delta_tree_test.go [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+			expectErr: assert.NoError,
+		},
+		{
+			tname:     "add folder",
+			tree:      treeWithRoot(),
+			parentID:  odConsts.RootID,
+			id:        "folder",
+			name:      "nameyMcNameFace",
+			expectErr: assert.NoError,
+		},
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/delta_tree_test.go:104]: I think this can be removed as it's a duplicate of the below test
```suggestion
```

File: src/internal/m365/collection/drive/delta_tree_test.go [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+
+			result := test.tree.folderIDToNode[test.id]
+			require.NotNil(t, result)
+			assert.Nil(t, result.prev)
+			assert.Equal(t, test.id, result.id)
+			assert.Equal(t, test.name, result.name)
+			assert.Equal(t, test.isPackage, result.isPackage)
+
+			_, ded := test.tree.tombstones[test.id]
+			assert.False(t, ded)
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/delta_tree_test.go:223]: nit: minor readability update
```suggestion
			assert.NotContains(t, test.tree.tombstones, test.id)
```

**Review by ashmrtn — APPROVED** [review]
Few minor nits but LGTM. We can deal with additional edge cases in new PRs as they come up

**Comment by ryanfkeepers** [comment]
@ashmrtn pushed a fairly significant rework to update this code to the latest graph behavior discoveries, and the assumption of a multi-enumeration backup.  Could I get a re-review, please?

File: src/internal/m365/collection/drive/delta_tree_test.go [hunk 11]
```diff
@@ -40,17 +41,663 @@ func (suite *DeltaTreeUnitSuite) TestNewNodeyMcNodeFace() {
...
+	set(lefty.id, "bloaty", "bl", false)
+
+	bloaty := tree.folderIDToNode["bloaty"]
+	an(
+		tree.root,
+		an(lefty, an(bloaty)),
+		an(righty),
+	).compare(t, tree, true)
+
+	// add a child to righty
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/delta_tree_test.go:486]: nit: comment consistency with the code
```suggestion
	// add another child to lefty
```

File: src/internal/m365/collection/drive/delta_tree_test.go [hunk 12]
```diff
@@ -40,17 +41,663 @@ func (suite *DeltaTreeUnitSuite) TestNewNodeyMcNodeFace() {
...
+		an(
+			lefty,
+			an(bloaty),
+			an(righty, an(brightly))),
+	).compare(t, tree, true)
+}
+
+// this test focuses on whether the tree is correct when bouncing back and forth
+// between live and tombstoned states on the same folder
+func (suite *DeltaTreeUnitSuite) TestFolderyMcFolderFace_SetFolder_correctTombstones() {
```
Comment (ashmrtn) [inline src/internal/m365/collection/drive/delta_tree_test.go:519]: nice test!

**Review by ashmrtn — APPROVED** [review]
Just nits here and there

File: src/internal/m365/collection/drive/delta_tree.go [hunk 13]
```diff
@@ -65,16 +71,179 @@ type nodeyMcNodeFace struct {
...
+
+	if len(loc) == 0 {
+		return clues.NewWC(ctx, "missing tombstone location")
+	}
+
+	// since we run mutiple enumerations, it's possible to see a folder added on the
+	// first enumeration that then gets deleted on the next.  This means that the folder
+	// was deleted while the first enumeration was in flight, and will show up even if
+	// the folder gets restored after deltion.
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/delta_tree.go:205]: Oh no!  How will I stop the plagu?

File: src/internal/m365/collection/drive/delta_tree.go [hunk 14]
```diff
@@ -65,16 +71,179 @@ type nodeyMcNodeFace struct {
...
+	// need to ensure we have the minimum requirements met for adding a node.
+	if len(id) == 0 {
+		return clues.NewWC(ctx, "missing folder ID")
+	}
+
+	if len(name) == 0 {
+		return clues.NewWC(ctx, "missing folder name")
+	}
+
+	if len(parentID) == 0 && name != odConsts.RootPathDir {
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/m365/collection/drive/delta_tree.go:109]: absolutely
